### DISCUSSION TITLE
Absolute beginner - looking to build an electronic version of the Great Highland Bagpipe !!

### DISCUSSION THREAD

**Discussion by cws1745** [pschatzmann/arduino-audio-tools#971] [Q&A] [answered]
I am of somewhat advanced years with only a basic knowledge of electronics and Arduino programming - but negligible knowledge of how to create music 'electronically'.   I am no longer fit enough to play the real Great Highland Bagpipes and I want to create a portable alternative which creates sounds as close to the real thing as is possible. 
I know what the frequency is of each of the nine notes and the frequency of the Bass and Tenor drones.   I have also 'mapped' the finger combinations required to produce each of the notes and have tried and tested the capacitive touch pins on an ESP32 Wroom 32D Dev board - on the _**assumption**_ this board would be suitable for my needs.
However, I confess I have absolutely no idea how to progress with producing sounds!   I have spent many hours searching the web for basic tutorials on this subject, I have found how to create the odd 'beep' or 'chirp' using PWM but anything better than this and I'm lost with the concepts and terminology i.e. it goes from real basics to rocket science with little in between!
I'd greatly appreciate any help in gaining, if nothing else, a high level understanding of how a realistic musical note can be produced from what is effectively a button press, if some form of 'sampling' needs to be done and stored in memory or is some form of synthesis required, is it by hardware or software?   I've seen references to 'Mozzi' - looked at the documentation but find it very confusing.
Clearly, I have a lot to learn - perhaps biting off more than I can chew - but I have a real need for the end product.    I'm keen to learn and sincerely hope that this community can point me in the right direction, identify relevsnt tutorials, books etc or for that matter tell me to stop wasting my time if it is believed that my aspirations are too great because I've left it too late in life to venture into this complex field.
I look forward to your responses.

Charles

**Comment by pschatzmann** [comment, marked as answer]
I did not have any background in audio. After my retirement I was picking up Microcontrollers as hoppy and thought that doing audio on these devices might be cool.  To implement a good sounding instrument is indeed quite a big task for a beginner: 

_How to eat an elephant ?
In small pieces!_

I don't know anything about Bagpipes, so I try to provide an answer how I would approach the topic for a generic instrument and let you decide what makes sense.

### Basic Functionality
- I would start with a recording of individual tones of a specific instrument. This gives the advantage that you can choose your favourite instrument, or even gives you the option support more than one instrument. Audacity might be a good free tool. Also read [this Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/MemoryStream:-Converting-a-File-to-Flash-Memory) 
- You would play a tone when a button is pressed and stop when it is released. This [example](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-memory_raw-i2s) would be your starting point. Just replace starwars with your recorded audio and set it to loop forever.
- You need to add the logic to be able to play multiple tones at the same time. In the AudioTools this topic is called mixing. I guess it will take some effort to get this right. This [example](https://github.com/pschatzmann/arduino-audio-tools/blob/c5e38e96ab26788f1c1623eb3e25fc12fe87ae67/examples/examples-audiokit/streams-memory_pcm-mixer-audiokit/streams-memory_pcm-mixer-audiokit.ino#L26) might guide you. In Mozzi you would just calculate the weighted sum.

### Improving the quality
- Sometimes the beginning and the end of a tone has some distinctive sound pattern. You can try to model this by splitting the recording into beginning, the middle which can be repeated endlessly and the end. Alternatively you can try to apply some kind of ADSR envelope on the recording from the first step.
- On some instruments the tone depends on the force that is applied. This can be modelled by applying a different volume or by using different recordings. My gut feeling is, that this might not be relevant for bagpipes.

### Optimisations
- By decreasing the sample rate you reduce the needed memory to store your recordings. You can also use 8 bits instead of 16 or use some encoded audio to reduce it even further.
-  You can try to find out if you can just use the recording of 1 note and generate the other notes dynamically by resampling or with the help of pitch shifting.
- You can try to add midi support to your sketch, so that you can trigger notes e.g. from a keyboard

This suggestion should be valid whatever tool you decide to use:  Be it [Mozzi](https://sensorium.github.io/Mozzi/), [Maximilian](https://www.youtube.com/watch?v=H-Av78mtFF4&themeRefresh=1) or my AudioTools.  

You might also want to have a look at the available Arduino Bagpipe projects.

I think that the ESP32 is a good platform with a lot of Program Memory that you might consider as your first choice to store your recordings and the I2S interface together with an external DAC guarantees some good quality sound.

**Comment by cws1745** [comment]
Many thanks.   A lot to learn – taking my first bite by looking at Audacity and how to create a sound sample.

 

Best

 

Charles
